File: src/429. N-ary Tree Level Order Traversal/main.ts
// Solved by @kitanoyoru
// https://leetcode.com/problems/n-ary-tree-level-order-traversal/

/**
 * Definition for node.
 * class Node {
 *     val: number
 *     children: Node[]
 *     constructor(val?: number) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.children = []
 *     }
 * }
 */

const levelOrder = (root: Node | null): number[][] => {
  if (!root) {
    return []
  }
  let arr: number[][] = []

  let q: Node[] = []
  q.push(root)

  while (q.length) {
    let level: number[] = []
    let size = q.length
    for (let i = 0; i < size; i++) {
      let n = q.shift()!
      level.push(n.val)
      for (let u of n.children) {
        q.push(u)
      }
    }
    arr.push(level)
  }

  return arr
}

File: src/429. N-ary Tree Level Order Traversal/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/n-ary-tree-level-order-traversal/

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
  public List<List<Integer>> levelOrder(Node root) {
    if (root == null) {
      return new ArrayList<>();
    }
    List<List<Integer>> ans = new ArrayList<>();
    
    Queue<Node> q = new LinkedList<>();
    q.add(root);

    while (!q.isEmpty()) {
      int len = q.size();
      List<Integer> level = new ArrayList<>();
      
      while (len-- != 0) {
        Node node = q.remove();

        level.add(node.val);
        for (Node child : node.children) {
          q.add(child);
        }
      }

      ans.add(level);
    }

    return ans;
  }
}

